crates/factor-outbound-pg/src/client.rs

```diff
@@ -0,0 +1,287 @@
+use tokio_postgres::{config::SslMode, types::ToSql, Row};
+use tokio_postgres::{Client as TokioClient, NoTls, Socket};
+
+#[async_trait]
+pub trait Client {
+    async fn build_client(address: &str) -> Result<Self>
+    where
+        Self: Sized;
+
+    async fn execute(
+        &self,
+        statement: String,
+        params: Vec<ParameterValue>,
+    ) -> Result<u64, v2::Error>;
+
+    async fn query(
+        &self,
+        statement: String,
+        params: Vec<ParameterValue>,
+    ) -> Result<RowSet, v2::Error>;
+}
+
+#[async_trait]
+impl Client for TokioClient {
+    async fn build_client(address: &str) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        let config = address.parse::<tokio_postgres::Config>()?;
+
+        tracing::debug!("Build new connection: {}", address);
+
+        async move {
+            if config.get_ssl_mode() == SslMode::Disable {
+                let (client, connection) = config.connect(NoTls).await?;
+                spawn(connection);
+                Ok(client)
+            } else {
+                let builder = TlsConnector::builder();
+                let connector = MakeTlsConnector::new(builder.build()?);
+                let (client, connection) = config.connect(connector).await?;
+                spawn(connection);
+                Ok(client)
+            }
+        }
+        .await
+    }
+
+    async fn execute(
+        &self,
+        statement: String,
+        params: Vec<ParameterValue>,
+    ) -> Result<u64, v2::Error> {
+        let params: Vec<&(dyn ToSql + Sync)> = params
+            .iter()
+            .map(to_sql_parameter)
+            .collect::<Result<Vec<_>>>()
+            .map_err(|e| v2::Error::ValueConversionFailed(format!("{:?}", e)))?;
+
+        self.execute(&statement, params.as_slice())
+            .await
+            .map_err(|e| v2::Error::QueryFailed(format!("{:?}", e)))
+    }
+
+    async fn query(
+        &self,
+        statement: String,
+        params: Vec<ParameterValue>,
+    ) -> Result<RowSet, v2::Error> {
+        let params: Vec<&(dyn ToSql + Sync)> = params
+            .iter()
+            .map(to_sql_parameter)
+            .collect::<Result<Vec<_>>>()
+            .map_err(|e| v2::Error::BadParameter(format!("{:?}", e)))?;
+
+        let results = self
+            .query(&statement, params.as_slice())
+            .await
+            .map_err(|e| v2::Error::QueryFailed(format!("{:?}", e)))?;
+
+        if results.is_empty() {
+            return Ok(RowSet {
+                columns: vec![],
+                rows: vec![],
+            });
+        }
+
+        let columns = infer_columns(&results[0]);
+        let rows = results
+            .iter()
+            .map(convert_row)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|e| v2::Error::QueryFailed(format!("{:?}", e)))?;
+
+        Ok(RowSet { columns, rows })
+    }
+}
+
+fn spawn<T>(connection: tokio_postgres::Connection<Socket, T>)
```

Comment:
I was confused for a minute thinking this was just `use tokio::spawn`; might help to give it a distinct name:

```suggestion
fn spawn_connection<T>(connection: tokio_postgres::Connection<Socket, T>)
```